File: go/test/package_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestPackageEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Package(nil)
		if ent == nil {
			t.Fatal("expected non-nil PackageEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"package": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Package(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Package(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := packageBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "package." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PACKAGE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		packageRef01Ent := client.Package(nil)
		packageRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "package"}), "package_ref01"))
		packageRef01Data["org"] = setup.idmap["org01"]
		packageRef01Data["org_id"] = setup.idmap["org01"]
		packageRef01Data["package_id"] = setup.idmap["package01"]
		packageRef01Data["package_name"] = setup.idmap["package_name01"]
		packageRef01Data["package_type"] = setup.idmap["package_type01"]
		packageRef01Data["username"] = setup.idmap["username01"]
		packageRef01Data["version_id"] = setup.idmap["version01"]

		packageRef01DataResult, err := packageRef01Ent.Create(packageRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		packageRef01Data = core.ToMapAny(entityData(packageRef01DataResult))
		if packageRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if packageRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		packageRef01Match := map[string]any{}

		packageRef01ListResult, err := packageRef01Ent.List(packageRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		packageRef01List, packageRef01ListOk := packageRef01ListResult.([]any)
		if !packageRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", packageRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(packageRef01List), map[string]any{"id": packageRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		packageRef01MatchDt0 := map[string]any{
			"id": packageRef01Data["id"],
		}
		packageRef01DataDt0Loaded, err := packageRef01Ent.Load(packageRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		packageRef01DataDt0LoadResult := core.ToMapAny(entityData(packageRef01DataDt0Loaded))
		if packageRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if packageRef01DataDt0LoadResult["id"] != packageRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		packageRef01MatchRm0 := map[string]any{
			"id": packageRef01Data["id"],
		}
		_, err = packageRef01Ent.Remove(packageRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		packageRef01MatchRt0 := map[string]any{}

		packageRef01ListRt0Result, err := packageRef01Ent.List(packageRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		packageRef01ListRt0, packageRef01ListRt0Ok := packageRef01ListRt0Result.([]any)
		if !packageRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", packageRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(packageRef01ListRt0), map[string]any{"id": packageRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func packageBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "package", "PackageTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read package test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse package test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"package01", "package02", "package03", "org01", "org02", "org03", "user01", "user02", "user03", "version01", "version02", "version03", "package_name01", "package_type01", "username01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_PACKAGE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_PACKAGE_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_PACKAGE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/PackageEntityTest.php
<?php
declare(strict_types=1);

// Package entity test

require_once __DIR__ . '/../github_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class PackageEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GithubSDK::test(null, null);
        $ent = $testsdk->Package(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "package" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = GithubSDK::test($seed, null);
        $seen = iterator_to_array($base->Package(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = GithubConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = GithubSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Package(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = package_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "package." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PACKAGE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $package_ref01_ent = $client->Package(null);
        $package_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.package"), "package_ref01"));
        $package_ref01_data["org"] = $setup["idmap"]["org01"];
        $package_ref01_data["org_id"] = $setup["idmap"]["org01"];
        $package_ref01_data["package_id"] = $setup["idmap"]["package01"];
        $package_ref01_data["package_name"] = $setup["idmap"]["package_name01"];
        $package_ref01_data["package_type"] = $setup["idmap"]["package_type01"];
        $package_ref01_data["username"] = $setup["idmap"]["username01"];
        $package_ref01_data["version_id"] = $setup["idmap"]["version01"];

        $package_ref01_data_result = $package_ref01_ent->create($package_ref01_data, null);
        $package_ref01_data = Helpers::to_map(is_object($package_ref01_data_result) && method_exists($package_ref01_data_result, 'data_get') ? $package_ref01_data_result->data_get() : $package_ref01_data_result);
        $this->assertNotNull($package_ref01_data);
        $this->assertNotNull($package_ref01_data["id"]);

        // LIST
        $package_ref01_match = [];

        $package_ref01_list_result = $package_ref01_ent->list($package_ref01_match, null);
        $this->assertIsArray($package_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($package_ref01_list_result),
            ["id" => $package_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // LOAD
        $package_ref01_match_dt0 = [
            "id" => $package_ref01_data["id"],
        ];
        $package_ref01_data_dt0_loaded = $package_ref01_ent->load($package_ref01_match_dt0, null);
        $package_ref01_data_dt0_load_result = Helpers::to_map(is_object($package_ref01_data_dt0_loaded) && method_exists($package_ref01_data_dt0_loaded, 'data_get') ? $package_ref01_data_dt0_loaded->data_get() : $package_ref01_data_dt0_loaded);
        $this->assertNotNull($package_ref01_data_dt0_load_result);
        $this->assertEquals($package_ref01_data_dt0_load_result["id"], $package_ref01_data["id"]);

        // REMOVE
        $package_ref01_match_rm0 = [
            "id" => $package_ref01_data["id"],
        ];
        $package_ref01_ent->remove($package_ref01_match_rm0, null);

        // LIST
        $package_ref01_match_rt0 = [];

        $package_ref01_list_rt0_result = $package_ref01_ent->list($package_ref01_match_rt0, null);
        $this->assertIsArray($package_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($package_ref01_list_rt0_result),
            ["id" => $package_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function package_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/package/PackageTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GithubSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["package01", "package02", "package03", "org01", "org02", "org03", "user01", "user02", "user03", "version01", "version02", "version03", "package_name01", "package_type01", "username01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GITHUB_TEST_PACKAGE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GITHUB_TEST_PACKAGE_ENTID" => $idmap,
        "GITHUB_TEST_LIVE" => "FALSE",
        "GITHUB_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GITHUB_TEST_PACKAGE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["GITHUB_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        // "?? []" because merge legitimately answers with a stdClass when every
        // contributing entry is an EMPTY map - an SDK with no apikey and no
        // server variables generates an empty middle entry, so that is the
        // common case, not the edge one. to_map returns null for a non-array by
        // design, and the constructor takes a non-nullable array, so without the
        // fallback every such SDK died on "must be of type array, null given"
        // the moment live mode was switched on. Offline mode never reaches this
        // branch, which is why the offline suite stayed green.
        $client = new GithubSDK(Helpers::to_map($merged_opts) ?? []);
    }

    $live = $env["GITHUB_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GITHUB_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
